src/ssh/allowed_signer.rs

```diff
@@ -0,0 +1,456 @@
+    pub(in self) fn next(&mut self, opening_quotes_allowed: bool) -> Result<Option<String>> {
+        if self.is_empty_after_trim() {
+            return Ok(None);
+        }
+
+        // If the next token starts with a double quote, then the closing double quote is also
+        // the end of the token
+        if self.buffer[0] == "\"" {
+            if opening_quotes_allowed {
+                return self.split_quote().map(|v| Some(v));
+            } else {
+                return Err(Error::InvalidAllowedSigner(AllowedSignerParsingError::InvalidQuotes));
+            }
+        }
+
+        // If the next token doesn't start with a double quote, the token can represent an option.
+        // Only an option token can contain double quotes in the middle (e.g. a="b c").
+        // If we don't see any double quote in the token, we greedily parse the token until the
+        // next whitespace.
+        let mut s = String::new();
+        while !self.buffer.is_empty()
+            && ![" ", "\"", "#"].contains(&self.buffer.last().unwrap().as_str()) {
+            s.push_str(&self.buffer.pop().unwrap());
+        }
+
+        // This should only apply to options
+        if !self.buffer.is_empty() && self.buffer.last().unwrap().as_str() == "\"" {
+            s.push_str(self.split_quote()?.as_str());
+
+            // After the double quotes in the option token, there can only be nothing, a
+            // whitespace, or a pound
+            if !self.buffer.is_empty()
+                && ![" ", "#"].contains(&self.buffer.last().unwrap().as_str()) {
+                return Err(Error::InvalidAllowedSigner(AllowedSignerParsingError::InvalidQuotes));
+            }
+        }
+
+        Ok(Some(s))
+    }
+
+    /// Trim comment and whitespaces
+    fn trim(&mut self) {
+        while !self.buffer.is_empty(){
+            match self.buffer.last().unwrap().as_str() {
+                " " => {
+                    self.buffer.pop();
+                },
+                // Comment detected
+                "#" => {
+                    self.buffer.clear()
+                },
+                _ => break,
+            };
+        }
+    }
+
+    /// Extract content inside the double quotes.
+    /// This function assumes buffer starst with a ".
+    fn split_quote(&mut self) -> Result<String> {
+        if self.buffer.is_empty() || self.buffer.pop().unwrap() != "\"" {
+            return Err(Error::InvalidAllowedSigner(AllowedSignerParsingError::InvalidQuotes));
+        }
+
+        let mut s = String::from("\"");
+        loop {
+            let token = self.buffer.pop()
+                .ok_or(Error::InvalidAllowedSigner(AllowedSignerParsingError::InvalidQuotes))?;
+            s.push_str(&token);
+            if token == "\"" {
+                break;
+            }
+        }
+
+        Ok(s)
+    }
+}
+
+/// Parse a string into a u64 representing a timestamp.
+/// The timestamp has format YYYYMMDD[HHMM[SS]][Z]
```

Comment:
I can't believe this isn't a defined time format like RFC3339 or ISO 8601

Reply:
Me sad